Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift

```diff
@@ -15,38 +15,92 @@ import Dispatch
 
 /// The solver that is able to transitively resolve a set of package constraints
 /// specified by a root package.
-public final class PubgrubDependencyResolver {
-
+public struct PubgrubDependencyResolver {
     /// The type of the constraints the resolver operates on.
     public typealias Constraint = PackageContainerConstraint
 
-    /// The current best guess for a solution satisfying all requirements.
-    public var solution = PartialSolution()
-
-    /// A collection of all known incompatibilities matched to the packages they
-    /// refer to. This means an incompatibility can occur several times.
-    public var incompatibilities: [DependencyResolutionNode: [Incompatibility]] = [:]
+    /// the mutable state that get computed
+    internal final class State {
+        /// The root package reference.
+        let root: DependencyResolutionNode
+
+        /// The list of packages that are overridden in the graph. A local package reference will
+        /// always override any other kind of package reference and branch-based reference will override
+        /// version-based reference.
+        let overriddenPackages: [PackageReference: (version: BoundVersion, products: ProductFilter)]
+
+        /// A collection of all known incompatibilities matched to the packages they
+        /// refer to. This means an incompatibility can occur several times.
+        public private(set) var incompatibilities: [DependencyResolutionNode: [Incompatibility]] = [:]
+
+        /// The current best guess for a solution satisfying all requirements.
+        public private(set) var solution: PartialSolution
+        
+        private let lock = Lock()
+
+        init(root: DependencyResolutionNode,
+             overriddenPackages: [PackageReference: (version: BoundVersion, products: ProductFilter)] = [:],
+             solution: PartialSolution = PartialSolution()) {
+            self.root = root
+            self.overriddenPackages = overriddenPackages
+            self.solution = solution
+        }
 
-    /// Find all incompatibilities containing a positive term for a given package.
-    public func positiveIncompatibilities(for node: DependencyResolutionNode) -> [Incompatibility]? {
-        guard let all = incompatibilities[node] else {
-            return nil
+        func addIncompatibility(_ incompatibility: Incompatibility, at location: LogLocation) {
+            self.lock.withLock {
+                //log("incompat: \(incompatibility) \(location)")
```

Comment:
remove?

Reply:
trying to minimize changes for now, we can revisit in later PR